Python/Product/Analysis/LanguageServer/Server.cs

```diff
@@ -214,46 +216,46 @@ private sealed class ReloadModulesQueueItem : IAnalyzable {
         }
 
         public override Task DidCloseTextDocument(DidCloseTextDocumentParams @params) {
-            var doc = _projectFiles.GetEntry(@params.textDocument.uri) as IDocument;
+            _editorFiles.Close(@params.textDocument.uri);
 
+            var doc = ProjectFiles.GetEntry(@params.textDocument.uri) as IDocument;
             if (doc != null) {
                 // No need to keep in-memory buffers now
                 doc.ResetDocument(-1, null);
-
-                _openFiles.Remove(@params.textDocument.uri);
-
                 // Pick up any changes on disk that we didn't know about
                 EnqueueItem(doc, AnalysisPriority.Low);
             }
+
             return Task.CompletedTask;
         }
 
 
         public override async Task DidChangeConfiguration(DidChangeConfigurationParams @params) {
-            if (_analyzer == null) {
+            if (Analyzer == null) {
                 LogMessage(MessageType.Error, "change configuration notification sent to uninitialized server");
                 return;
             }
 
-            if (@params.settings == null) {
-                return;
-            }
-
-            if (@params.settings is LanguageServerSettings settings) {
-                _settings = settings;
-            } else {
-                LogMessage(MessageType.Error, "change configuration notification sent unsupported settings");
-                return;
+            var reanalyze = true;
+            if (@params.settings != null) {
+                if (@params.settings is LanguageServerSettings settings) {
+                    reanalyze = HandleConfigurationChanges(settings);
+                } else {
+                    LogMessage(MessageType.Error, "change configuration notification sent unsupported settings");
+                    return;
+                }
             }
 
-            // Make sure reload modules is executed on the analyzer thread.
-            var task = _reloadModulesQueueItem.Task;
-            _queue.Enqueue(_reloadModulesQueueItem, AnalysisPriority.Normal);
-            await task;
+            if (reanalyze) {
+                // Make sure reload modules is executed on the analyzer thread.
+                var task = _reloadModulesQueueItem.Task;
+                _queue.Enqueue(_reloadModulesQueueItem, AnalysisPriority.Normal);
+                await task;
 
-            // re-analyze all of the modules when we get a new set of modules loaded...
-            foreach (var entry in _analyzer.ModulesByFilename) {
-                _queue.Enqueue(entry.Value.ProjectEntry, AnalysisPriority.Normal);
+                // re-analyze all of the modules when we get a new set of modules loaded...
+                foreach (var entry in Analyzer.ModulesByFilename) {
```

Comment:
`Analyzer` may become null here because of the `Shutdown` concurrent call. We should either check it for null again or cache the value of the field at the beginning

Reply:
Or we can catch NRE and throw `OperationCanceledException` instead

Reply:
It shouldn't ever be concurrent, but I vote for either caching it or just not setting it to null and making it do ObjectDisposedException properly.

Reply:
Actually, looking at the broader scope, that may happen here.

Reply:
Analyzer is pretty big to track disposed state atm... Ditto for `AnalysisQueue`. We may need to port `DisposableBag` here. I'll just check it here. The condition is not particularly critical since if server is shutdown, the worst thing is that it continues consuming some CPU for few seconds. Also, we are in a single thread context here so I am not sure Shutdown can come while `DidChangeConfiguration` is still running.

Reply:
`ObjectDisposedException` should be thrown only if we want calling code to be able to restart/reinitialize disposed object. If we simply want to cancel the execution, it is better to throw `OperationCanceledException`. We have another disposable primitive, `DisposeToken` (https://github.com/Microsoft/RTVS/blob/master/src/Common/Core/Impl/Disposables/DisposeToken.cs), which can be used in async methods.

Reply:
I'd do OperationCanceledException since it is generally normal condition.

Reply:
We previously awaited on a task running in the analysis thread, so I think we could end up having Shutdown come in between.

IMO, ObjectDisposedException is a perfectly good signal that we should abandon the object. In this context, that means exit the process, since we only have one server per process.

Reply:
OperationCanceledException may already be used from this class to indicate that you can send another request, so it becomes overloaded on context if we use it here because the server was shut down.

Reply:
Would it cause unnecessary Watsons? I.e. analysis still running when host exits?

Reply:
But you don't abandon the object. You abandon the code flow. In rare case where you care why it is abandoned you need to ask the object directly, cause `ObjectDisposedException` doesn't show you what object is disposed (it can be server, it can be stream, it can be some local scope). But in most cases you just want to stop execution.

Reply:
I don't want to add try/catch blocks to every method that uses server. This is pretty much the same trash code that handles error codes.

Reply:
Except the reason this error occurs is because `Shutdown` was called and the server should be abandoned.

If we were handling request cancellation here, then yeah, we just want to cancel the operation. But when it's because there is no point using this server ever again I think ObjectDisposedException is totally appropriate (even if it's being thrown somewhat asynchronously).

@MikhailArkhipov If the call that's awaiting on this is handling whatever exception we throw, then there shouldn't be any crash reports. The callback shouldn't go away, though it likely will fail to send a response at that point (though I don't think it has a response to send here anyway).

Reply:
> Except the reason this error occurs is because `Shutdown` was called and the server should be abandoned.

How this case is different? Shutdown is called only through JsonRPC. There will be no more new requests after that, only those that are currently unfinished. Client code that triggered them doesn't care why they aren't completed and it shouldn't handle server disposed state. Its the job of the code that called Shutdown or the code that is going to send new request (spin the new instance of the process first).

Reply:
Then we shouldn't throw any exception from here at all and should silently exit.

Though I'd rather do that by either checking a flag on the server, on the analyzer, or handling an ObjectDisposedException here and using that to silently stop (as we do throughout the debugger, for example).

Reply:
> Then we shouldn't throw any exception from here at all and should silently exit

Exactly, so that is what `OperationCanceledException` does. C# doesn't have any language-specific mechanism that allows us to abort some code flow, it uses OCE instead. 

`OperationDisposedException` is very old, it was added to the FCL when C# code was totally thread-bound, and `IDisposable` intended to serve completely different purpose (from which we still have that absurd FxCop requirement about full `IDisposable` implementation).

Reply:
Soooo, `OperationCanceledException` it is?

Reply:
Apparently, this discussion is mostly academical :-) in VS Code we get 'Shutdown' which, per LSP spec should NOT exit, but never get Exit since right after the Shutdown, without any waits, the process is terminated. Happens when either VSC exits or workspace is closed. 

So, the process is insta-gone, it is pretty low probability it will be able to do much after the shutdown request.